include/maths/CLocalOutlierFactors.h

```diff
@@ -184,6 +190,47 @@ class MATHS_EXPORT CLocalOutlierFactors {
         }
     }
 
+    //! Estimate the amount of memory that will be used computing outliers.
+    //!
+    //! \param[in] algorithm The algorithm that will be used.
+    //! \param[in] k The number of nearest neighbours which will be used.
+    //! \param[in] numberPoints The number of points for outliers will be
+    //! computed.
+    //! \param[in] dimension The dimension of the points for which outliers
+    //! will be computed.
+    template<typename POINT>
+    static std::size_t estimateMemoryUsage(EAlgorithm algorithm,
+                                           std::size_t k,
+                                           std::size_t numberPoints,
+                                           std::size_t dimension) {
+        std::size_t result{TKdTree<POINT>::estimateMemoryUsage(numberPoints, dimension)};
+        switch (algorithm) {
+        case E_Ensemble:
+            result += CLof<POINT, TKdTree<POINT>>::estimateMemoryUsage(k, numberPoints) +
+                      CEnsemble<POINT, TKdTree<POINT>>::estimateMemoryUsage(
+                          numberPoints, dimension);
```

Comment:
Since Ensemble could include `lof` why do we add the size of `lof` here as well? Does ensemble not care?

Reply:
As it stands each class only accounts for their own memory usage in estimateMemoryUsage, so `CEnsemble` is only accounting for its own state and not that of `CLof`.

There is a bit of a rough edge here in that the ensemble can have any collection of methods added to it at run time, but this is not configured anywhere, and this code just assumes that all of them will be included.

So, the reason I did this rather than have CEnsemble::estimateMemoryUsage account for the memory usage of the underlying methods as well, was that I was thinking down the line, it might be cleaner to use a bit mask for selecting subsets of the methods to use as an ensemble. In this case, we'd change the switch to something like:
```
if (algorithm & E_Lof) {
}
...
```
and it seemed cleaner to then have each class **only** account for own memory overheads.

Given the confusion, I think it is fair to say that I should make the naming of this method clearer; how about `estimateMemoryDirectlyUsed`?

Reply:
Ah, I see now. Yeah, a rename would be nice as `estimateMemoryUsage` does seem to imply "The memory I am using" would could easily encompass the underlying methodologies.

Reply:
[972d15b](https://github.com/elastic/ml-cpp/pull/345/commits/972d15b0e70e9742b314c15133beb59bfb9c8713).

This also fixes a couple of silly bugs. These would get caught when we can test running memory usage vs estimated memory usage, which needs the monitoring framework in place.